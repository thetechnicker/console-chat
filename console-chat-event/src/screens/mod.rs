use crate::event::WidgetEvent;
use ratatui::layout::Rect;
use std::fmt::Debug;

#[derive(Default, Debug, Clone)]
pub enum CurrentScreen {
    #[default]
    Login,
    Chat,
}

pub struct ScreenState {
    pub title: String,
    pub hint_area: Rect,
}

pub trait Screen: Debug {
    fn handle_event(&mut self, event: WidgetEvent);
    //fn render(&self, area: Rect, buf: &mut Buffer);
}

pub mod chat;
pub use chat::*;
pub mod login;
pub use login::*;
