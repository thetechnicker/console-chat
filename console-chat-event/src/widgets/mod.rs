use crate::event::WidgetEvent;
//use ratatui::widgets as ratatui_widgets;
//use ratatui::crossterm::event::KeyEvent;
use ratatui::{buffer::Buffer, layout::Rect};
use std::fmt::Debug;

pub trait Widget: Debug {
    fn handle_event(&mut self, event: WidgetEvent);
    fn draw(&self, area: Rect, buf: &mut Buffer);
}

pub mod input;
pub use input::*;
pub mod button;
pub use button::*;
