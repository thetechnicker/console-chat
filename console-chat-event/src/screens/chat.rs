use crate::DEFAULT_BORDER;
use crate::event::{AppEvent, EventSender};
use crate::screens::{CursorPos, Screen};
use crate::widgets;
use crate::widgets::Widget;
use crossterm::event::{KeyCode, KeyEventKind};
use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Layout, Rect},
    widgets::{Block, Paragraph, Widget as UiWidget},
};

#[derive(Debug)]
pub struct ChatScreen {
    pub tab_index: usize,
    pub max_tab: usize,
    pub event_sender: EventSender,
    pub input: widgets::InputWidget,
}

impl ChatScreen {
    pub fn new(event_sender: EventSender) -> Self {
        Self {
            tab_index: 0,
            max_tab: 2,
            event_sender,
            input: widgets::InputWidget::default(),
        }
    }
    pub fn send_current_widget_event(&mut self, event: AppEvent) {
        if let Some(elem) = self.current_widget() {
            elem.handle_event(event)
        }
    }
    pub fn send_all_widgets_event(&mut self, event: AppEvent) {
        for i in 0..self.max_tab {
            if let Some(elem) = self.widget_at(i) {
                elem.handle_event(event.clone());
            }
        }
    }

    pub fn widget_at(&mut self, index: usize) -> Option<&mut dyn Widget> {
        match index {
            1 => Some(&mut self.input as &mut dyn Widget),
            _ => None,
        }
    }
    pub fn current_widget(&mut self) -> Option<&mut dyn Widget> {
        match self.tab_index {
            1 => Some(&mut self.input as &mut dyn Widget),
            _ => None,
        }
    }
}

impl Screen for ChatScreen {
    fn handle_event(&mut self, event: AppEvent) {
        match event {
            AppEvent::Clear(hard) => {
                self.tab_index = 0;
                for i in 0..self.max_tab {
                    if let Some(w) = self.widget_at(i) {
                        w.handle_event(AppEvent::Clear(hard));
                    }
                }
            }
            AppEvent::KeyEvent(key_event) => {
                match key_event.code {
                    KeyCode::Tab if key_event.kind == KeyEventKind::Press => {
                        self.send_current_widget_event(AppEvent::NoFocus);
                        self.tab_index = (self.tab_index + 1) % self.max_tab;
                        self.send_current_widget_event(AppEvent::Focus);
                    }
                    KeyCode::BackTab if key_event.kind == KeyEventKind::Press => {
                        self.send_current_widget_event(AppEvent::NoFocus);
                        self.tab_index = (self.tab_index - 1) % self.max_tab;
                        self.send_current_widget_event(AppEvent::Focus);
                    }
                    KeyCode::Esc => {
                        self.send_all_widgets_event(AppEvent::NoFocus);
                        self.tab_index = 0;
                    }

                    _ => {}
                }
                self.send_current_widget_event(AppEvent::KeyEvent(key_event));
            }
            _ => {}
        };
    }
    /*
    }

    impl UiWidget for &ChatScreen {
        fn render(self, area: Rect, buf: &mut Buffer) {
        */
    fn draw(&self, area: Rect, buf: &mut Buffer) -> Option<CursorPos> {
        // MAIN
        let chat_block = Block::bordered().border_type(DEFAULT_BORDER);
        let chat_inner = chat_block.inner(area);
        chat_block.render(area, buf);

        let [chat, input] =
            Layout::vertical([Constraint::Min(10), Constraint::Max(3)]).areas(chat_inner);

        let x = Paragraph::new(format!("{:?}\n{}", self.input, self.tab_index));
        x.render(chat, buf);

        // Input
        self.input.draw(input, buf, &mut None);
        None
    }
}
