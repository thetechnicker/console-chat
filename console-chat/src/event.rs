use crate::network::NetworkEvent;
use crate::screens::CurrentScreen;
use color_eyre::eyre::OptionExt;
use crossterm::event::{Event as CrosstermEvent, KeyCode, KeyEvent, KeyModifiers};
use futures::{FutureExt, StreamExt};
use std::time::Duration;
use tokio::sync::mpsc;

/// The frequency at which tick events are emitted.
const TICK_FPS: f64 = 30.0; // 1.0;

/// Representation of all possible events.
#[derive(Clone, Debug)]
pub enum Event {
    /// An event that is emitted on a regular schedule.
    ///
    /// Use this event to run any code which has to run outside of being a direct response to a user
    /// event. e.g. polling exernal systems, updating animations, or rendering the UI based on a
    /// fixed frame rate.
    Tick,
    /// Crossterm events.
    ///
    /// These events are emitted by the terminal.
    Crossterm(CrosstermEvent),
    /// Application events.
    ///
    /// Use this event to emit custom events that are specific to your application.
    App(AppEvent),
}

impl From<AppEvent> for Event {
    fn from(app_event: AppEvent) -> Self {
        Self::App(app_event)
    }
}

impl From<CrosstermEvent> for Event {
    fn from(event: CrosstermEvent) -> Self {
        Self::Crossterm(event)
    }
}

/*
#[derive(Clone, Debug)]
pub enum AppEvent {
    Focus,
    NoFocus,
    Clear,
    KeyEvent(KeyEvent),
}
*/

/// Application events.
///
/// You can extend this enum with your own custom events.
#[derive(Clone, Debug)]
pub enum AppEvent {
    /// Quit the application.
    Quit,

    Focus,
    NoFocus,
    Clear(bool),

    //TriggerApiReconnect,
    NetworkEvent(NetworkEvent),

    KeyEvent(KeyEvent),
    ButtonPress(String),
    SwitchScreen(CurrentScreen),
    SendMessage(String),
    SimpleMSG(String),
}

/// Terminal event handler.
#[derive(Debug)]
pub struct EventHandler {
    /// Event sender channel.
    sender: mpsc::UnboundedSender<Event>,
    /// Event receiver channel.
    receiver: mpsc::UnboundedReceiver<Event>,
    handle: tokio::task::JoinHandle<Result<(), color_eyre::eyre::Error>>,
}

impl Default for EventHandler {
    fn default() -> Self {
        Self::new()
    }
}

impl EventHandler {
    /// Constructs a new instance of [`EventHandler`] and spawns a new thread to handle events.
    pub fn new() -> Self {
        let (sender, receiver) = mpsc::unbounded_channel();
        let actor = EventTask::new(sender.clone());
        let handle = tokio::spawn(async { actor.run().await });
        Self {
            sender,
            receiver,
            handle,
        }
    }

    pub fn get_event_sender(&self) -> EventSender {
        EventSender::new(self.sender.clone())
    }

    pub fn stop(&self) {
        self.handle.abort();
    }

    /// Receives an event from the sender.
    ///
    /// This function blocks until an event is received.
    ///
    /// # Errors
    ///
    /// This function returns an error if the sender channel is disconnected. This can happen if an
    /// error occurs in the event thread. In practice, this should not happen unless there is a
    /// problem with the underlying terminal.
    pub async fn next(&mut self) -> color_eyre::Result<Event> {
        self.receiver
            .recv()
            .await
            .ok_or_eyre("Failed to receive event")
    }

    /// Queue an app event to be sent to the event receiver.
    ///
    /// This is useful for sending events to the event handler which will be processed by the next
    /// iteration of the application's event loop.
    pub fn send(&mut self, app_event: AppEvent) {
        // Ignore the result as the reciever cannot be dropped while this struct still has a
        // reference to it
        let _ = self.sender.send(Event::App(app_event));
    }
}
///Simple struct so Screens can send back events to app
#[derive(Debug, Clone)]
pub struct EventSender {
    sender: mpsc::UnboundedSender<Event>,
}

/// Used for testing only
impl Default for EventSender {
    fn default() -> Self {
        let (tx, _rx) = mpsc::unbounded_channel();
        Self { sender: tx }
    }
}

impl EventSender {
    pub fn new(sender: mpsc::UnboundedSender<Event>) -> Self {
        Self { sender }
    }
    pub fn send(&self, event: Event) {
        // Ignores the result because shutting down the app drops the receiver, which causes the send
        // operation to fail. This is expected behavior and should not panic.
        let _ = self.sender.send(event);
    }
}
impl From<EventSender> for AppEventSender {
    fn from(val: EventSender) -> Self {
        AppEventSender::new(val.sender)
    }
}

impl From<EventSender> for NetworkEventSender {
    fn from(val: EventSender) -> Self {
        NetworkEventSender::new(val.sender)
    }
}

#[derive(Debug, Clone)]
pub struct AppEventSender {
    sender: mpsc::UnboundedSender<Event>,
}

impl AppEventSender {
    pub fn new(sender: mpsc::UnboundedSender<Event>) -> Self {
        Self { sender }
    }
    pub fn send(&self, event: AppEvent) {
        // Ignores the result because shutting down the app drops the receiver, which causes the send
        // operation to fail. This is expected behavior and should not panic.
        let _ = self.sender.send(Event::App(event));
    }
}

#[derive(Debug, Clone)]
pub struct NetworkEventSender {
    sender: mpsc::UnboundedSender<Event>,
}

impl NetworkEventSender {
    pub fn new(sender: mpsc::UnboundedSender<Event>) -> Self {
        Self { sender }
    }
    pub fn send(&self, event: NetworkEvent) {
        // Ignores the result because shutting down the app drops the receiver, which causes the send
        // operation to fail. This is expected behavior and should not panic.
        let _ = self.sender.send(Event::App(AppEvent::NetworkEvent(event)));
    }
}

/// A thread that handles reading crossterm events and emitting tick events on a regular schedule.
struct EventTask {
    /// Event sender channel.
    sender: mpsc::UnboundedSender<Event>,
}

/*
async fn get_event() -> Option<CrosstermEvent> {
    if event::poll(Duration::from_millis(1000)).unwrap_or(false) {
        if let Ok(event) = event::read() {
            return Some(event);
        }
    }
    None
}
*/

impl EventTask {
    /// Constructs a new instance of [`EventThread`].
    fn new(sender: mpsc::UnboundedSender<Event>) -> Self {
        Self { sender }
    }

    /// Runs the event thread.
    ///
    /// This function emits tick events at a fixed rate and polls for crossterm events in between.
    async fn run(self) -> color_eyre::Result<()> {
        let tick_rate = Duration::from_secs_f64(1.0 / TICK_FPS);
        let mut reader = crossterm::event::EventStream::new();
        let mut tick = tokio::time::interval(tick_rate);
        loop {
            let tick_delay = tick.tick();
            //let crossterm_event = get_event();
            let crossterm_event = reader.next().fuse();
            tokio::select! {
              _ = self.sender.closed() => {
                break;
              }
              _ = tick_delay => {
                self.send(Event::Tick);
              }
              Some(Ok(evt)) = crossterm_event => match evt {
                    crossterm::event::Event::Key(key_event) => {
                        self.send(handle_key_events(key_event))
                    }
                    _ => self.send(Event::Crossterm(evt))
                },
            };
        }
        Ok(())
    }

    /// Sends an event to the receiver.
    fn send(&self, event: Event) {
        // Ignores the result because shutting down the app drops the receiver, which causes the send
        // operation to fail. This is expected behavior and should not panic.
        let _ = self.sender.send(event);
    }
}

#[inline]
fn handle_key_events(key_event: KeyEvent) -> Event {
    match key_event.code {
        KeyCode::Char('c' | 'C') if key_event.modifiers == KeyModifiers::CONTROL => {
            Event::App(AppEvent::Quit)
        }
        _ => Event::App(AppEvent::KeyEvent(key_event)),
    }
}
