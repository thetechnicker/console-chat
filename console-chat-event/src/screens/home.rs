use crate::DEFAULT_BORDER;
use crate::event::{AppEvent, EventSender};
use crate::screens::{CursorPos, Screen};
use crate::widgets;
use crate::widgets::Widget;
use crossterm::event::{KeyCode, KeyEventKind};
use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Layout, Rect},
    widgets::{Block, Widget as UiWidget},
};
use serde_json;

#[derive(Debug)]
pub struct HomeScreen {
    pub tab_index: usize,
    pub max_tab: usize,
    pub event_sender: EventSender,
    pub room_input: widgets::InputWidget,
    pub join_button: widgets::Button,
    pub logout_button: widgets::Button,
    pub exit_button: widgets::Button,
}

impl HomeScreen {
    pub fn new(event_sender: EventSender) -> Self {
        Self {
            tab_index: 0,
            max_tab: 5,
            event_sender: event_sender.clone(),
            room_input: widgets::InputWidget::new("Room"),
            join_button: widgets::Button::new(
                "Join Room",
                event_sender.clone(),
                AppEvent::ButtonPress("JOIN".to_string()),
            )
            .theme(widgets::GREEN),
            logout_button: widgets::Button::new(
                "Logout",
                event_sender.clone(),
                AppEvent::ButtonPress("LOGOUT".to_string()),
            )
            .theme(widgets::BLUE),
            exit_button: widgets::Button::new("Exit", event_sender.clone(), AppEvent::Quit)
                .theme(widgets::RED),
        }
    }
    pub fn send_current_widget_event(&mut self, event: AppEvent) {
        if let Some(elem) = self.current_widget() {
            elem.handle_event(event)
        }
    }
    pub fn send_all_widgets_event(&mut self, event: AppEvent) {
        for i in 0..self.max_tab {
            if let Some(elem) = self.widget_at(i) {
                elem.handle_event(event.clone());
            }
        }
    }

    pub fn widget_at(&mut self, index: usize) -> Option<&mut dyn Widget> {
        match index {
            1 => Some(&mut self.room_input as &mut dyn Widget),
            2 => Some(&mut self.join_button as &mut dyn Widget),
            3 => Some(&mut self.logout_button as &mut dyn Widget),
            4 => Some(&mut self.exit_button as &mut dyn Widget),
            _ => None,
        }
    }
    pub fn current_widget(&mut self) -> Option<&mut dyn Widget> {
        self.widget_at(self.tab_index)
    }
}

impl Screen for HomeScreen {
    fn handle_event(&mut self, event: AppEvent) {
        match event {
            AppEvent::Clear(hard) => {
                self.tab_index = 0;
                for i in 0..self.max_tab {
                    if let Some(w) = self.widget_at(i) {
                        w.handle_event(AppEvent::Clear(hard));
                    }
                }
            }
            AppEvent::KeyEvent(key_event) => match key_event.code {
                KeyCode::Tab if key_event.kind == KeyEventKind::Press => {
                    self.send_current_widget_event(AppEvent::NoFocus);
                    self.tab_index = (self.tab_index.wrapping_add(1)) % self.max_tab;
                    self.send_current_widget_event(AppEvent::Focus);
                }
                KeyCode::BackTab if key_event.kind == KeyEventKind::Press => {
                    self.send_current_widget_event(AppEvent::NoFocus);
                    self.tab_index = (self.tab_index.wrapping_sub(1)) % self.max_tab;
                    self.send_current_widget_event(AppEvent::Focus);
                }
                KeyCode::Esc => {
                    self.send_all_widgets_event(AppEvent::NoFocus);
                    self.tab_index = 0;
                }
                _ => {
                    self.send_current_widget_event(AppEvent::KeyEvent(key_event));
                }
            },
            _ => {}
        };
    }

    fn get_data(&self) -> serde_json::Value {
        serde_json::json!(self.room_input.get_content())
    }

    /*
    }

    impl UiWidget for &HomeScreen {
        fn render(self, area: Rect, buf: &mut Buffer) {
        */
    fn draw(&self, area: Rect, buf: &mut Buffer) -> Option<CursorPos> {
        // MAIN
        let login_block = Block::bordered().border_type(DEFAULT_BORDER);
        let login_inner = login_block.inner(area);

        login_block.render(area, buf);
        let [_, input_area, _] = Layout::horizontal([
            Constraint::Fill(1),
            Constraint::Percentage(70),
            Constraint::Fill(1),
        ])
        .areas(login_inner);

        // Input
        let [_, user_input, _, buttons1, buttons2, _] = Layout::vertical([
            Constraint::Fill(1),
            Constraint::Max(3),
            Constraint::Max(3),
            Constraint::Max(3),
            Constraint::Max(3),
            Constraint::Fill(1),
        ])
        .areas(input_area);

        // User Input
        self.room_input.draw(user_input, buf, &mut None);

        // Buttons
        let x = 50;
        let [join_area, logout_area] =
            Layout::horizontal([Constraint::Percentage(x), Constraint::Percentage(x)])
                .areas(buttons1);
        self.join_button.draw(join_area, buf, &mut None);
        self.logout_button.draw(logout_area, buf, &mut None);
        self.exit_button.draw(buttons2, buf, &mut None);
        None
    }
}
