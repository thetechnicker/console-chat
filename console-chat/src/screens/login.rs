use crate::DEFAULT_BORDER;
use crate::event::{AppEvent, EventSender};
use crate::screens::{CursorPos, Screen};
use crate::widgets;
use crate::widgets::Widget;
use crossterm::event::{KeyCode, KeyEventKind};
use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Layout, Rect},
    widgets::{Block, Widget as UiWidget},
};
use serde_json;

#[derive(Debug)]
pub struct LoginScreen {
    pub tab_index: usize,
    pub max_tab: usize,
    pub event_sender: EventSender,
    pub user_input: widgets::InputWidget,
    pub pwd_input: widgets::InputWidget,
    pub skip_button: widgets::Button,
    pub ok_button: widgets::Button,
    pub cancel_button: widgets::Button,
}

impl LoginScreen {
    pub fn new(event_sender: EventSender) -> Self {
        Self {
            tab_index: 0,
            max_tab: 6,
            event_sender: event_sender.clone(),
            user_input: widgets::InputWidget::new("Username"),
            pwd_input: widgets::InputWidget::new("Password").password(),
            skip_button: widgets::Button::new(
                "Anonym",
                event_sender.clone(),
                AppEvent::ButtonPress("LOGIN_ANONYM".to_string()),
            )
            .theme(widgets::GREEN),
            ok_button: widgets::Button::new(
                "Login",
                event_sender.clone(),
                AppEvent::ButtonPress("LOGIN".to_string()),
            )
            .theme(widgets::BLUE),
            cancel_button: widgets::Button::new("Exit", event_sender.clone(), AppEvent::Quit)
                .theme(widgets::RED),
        }
    }
    pub fn send_current_widget_event(&mut self, event: AppEvent) {
        if let Some(elem) = self.current_widget_mut() {
            elem.handle_event(event)
        }
    }
    pub fn send_all_widgets_event(&mut self, event: AppEvent) {
        for i in 0..self.max_tab {
            if let Some(elem) = self.widget_at_mut(i) {
                elem.handle_event(event.clone());
            }
        }
    }

    pub fn current_widget(&self) -> Option<&dyn Widget> {
        self.widget_at(self.tab_index)
    }
    pub fn widget_at(&self, index: usize) -> Option<&dyn Widget> {
        match index {
            1 => Some(&self.user_input as &dyn Widget),
            2 => Some(&self.pwd_input as &dyn Widget),
            3 => Some(&self.ok_button as &dyn Widget),
            4 => Some(&self.skip_button as &dyn Widget),
            5 => Some(&self.cancel_button as &dyn Widget),
            _ => None,
        }
    }
    pub fn widget_at_mut(&mut self, index: usize) -> Option<&mut dyn Widget> {
        match index {
            1 => Some(&mut self.user_input as &mut dyn Widget),
            2 => Some(&mut self.pwd_input as &mut dyn Widget),
            3 => Some(&mut self.ok_button as &mut dyn Widget),
            4 => Some(&mut self.skip_button as &mut dyn Widget),
            5 => Some(&mut self.cancel_button as &mut dyn Widget),
            _ => None,
        }
    }
    pub fn current_widget_mut(&mut self) -> Option<&mut dyn Widget> {
        self.widget_at_mut(self.tab_index)
    }
}

impl Screen for LoginScreen {
    fn handle_event(&mut self, event: AppEvent) {
        match event {
            AppEvent::Clear(hard) => {
                self.tab_index = 0;
                for i in 0..self.max_tab {
                    if let Some(w) = self.widget_at_mut(i) {
                        w.handle_event(AppEvent::Clear(hard));
                    }
                }
            }
            AppEvent::KeyEvent(key_event) => match key_event.code {
                KeyCode::Tab if key_event.kind == KeyEventKind::Press => {
                    self.send_current_widget_event(AppEvent::NoFocus);
                    self.tab_index = (self.tab_index.wrapping_add(1)) % self.max_tab;
                    self.send_current_widget_event(AppEvent::Focus);
                }
                KeyCode::BackTab if key_event.kind == KeyEventKind::Press => {
                    self.send_current_widget_event(AppEvent::NoFocus);
                    self.tab_index = (self.tab_index.wrapping_sub(1)) % self.max_tab;
                    self.send_current_widget_event(AppEvent::Focus);
                }
                KeyCode::Esc => {
                    self.send_all_widgets_event(AppEvent::NoFocus);
                    self.tab_index = 0;
                }
                _ => {
                    self.send_current_widget_event(AppEvent::KeyEvent(key_event));
                }
            },
            _ => {}
        };
    }

    /*
    }
    impl UiWidget for &LoginScreen {
        fn render(self, area: Rect, buf: &mut Buffer) {
        */
    fn draw(&self, area: Rect, buf: &mut Buffer) -> Option<CursorPos> {
        // MAIN
        let login_block = Block::bordered().border_type(DEFAULT_BORDER);
        let login_inner = login_block.inner(area);

        login_block.render(area, buf);
        let [_, input_area, _] = Layout::horizontal([
            Constraint::Fill(1),
            Constraint::Percentage(70),
            Constraint::Fill(1),
        ])
        .areas(login_inner);

        // Input
        let [_, user_input, pwd_input, buttons, idk, _] = Layout::vertical([
            Constraint::Fill(1),
            Constraint::Max(3),
            Constraint::Max(3),
            Constraint::Max(3),
            Constraint::Max(3),
            Constraint::Fill(1),
        ])
        .areas(input_area);

        let mut u_x: Option<u16> = None;
        let mut p_x: Option<u16> = None;

        // User Input
        self.user_input.draw(user_input, buf, &mut u_x);

        // Password Input
        self.pwd_input.draw(pwd_input, buf, &mut p_x);

        // Buttons
        let x = 50;
        let [ok_area, anonym_area] =
            Layout::horizontal([Constraint::Percentage(x), Constraint::Percentage(x)])
                .areas(buttons);
        self.ok_button.draw(ok_area, buf, &mut None);
        self.skip_button.draw(anonym_area, buf, &mut None);
        self.cancel_button.draw(idk, buf, &mut None);

        if let Some(x) = u_x {
            Some(CursorPos {
                x: x + user_input.x as u16,
                y: user_input.y + 1 as u16,
            })
        } else if let Some(x) = p_x {
            Some(CursorPos {
                x: x + pwd_input.x as u16,
                y: pwd_input.y + 1 as u16,
            })
        } else {
            None
        }
    }

    fn get_data(&self) -> serde_json::Value {
        serde_json::json!({
            "username":self.user_input.get_content(),
            "password":self.pwd_input.get_content(),
        })
    }
}
