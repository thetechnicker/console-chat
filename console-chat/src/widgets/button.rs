use crate::event::{AppEvent, EventSender};
use crate::widgets::Widget;
use crossterm::event::KeyEventKind;
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{Color, Style},
    text::Line,
    widgets::{Block, Widget as w},
};

#[derive(Debug, Clone, Copy)]
pub struct Theme {
    text: Color,
    background: Color,
    highlight: Color,
    shadow: Color,
}

pub const BLUE: Theme = Theme {
    text: Color::Rgb(16, 24, 48),
    background: Color::Rgb(48, 72, 144),
    highlight: Color::Rgb(64, 96, 192),
    shadow: Color::Rgb(32, 48, 96),
};

pub const RED: Theme = Theme {
    text: Color::Rgb(48, 16, 16),
    background: Color::Rgb(144, 48, 48),
    highlight: Color::Rgb(192, 64, 64),
    shadow: Color::Rgb(96, 32, 32),
};

pub const GREEN: Theme = Theme {
    text: Color::Rgb(16, 48, 16),
    background: Color::Rgb(48, 144, 48),
    highlight: Color::Rgb(64, 192, 64),
    shadow: Color::Rgb(32, 96, 32),
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ButtonState {
    Normal,
    Selected,
    Active,
}

#[derive(Debug, Clone)]
pub struct Button {
    state: ButtonState,
    label: String,
    theme: Theme,
    event_sender: EventSender,
    on_press: AppEvent,
}

impl Button {
    pub fn new(label: &str, event_sender: EventSender, on_press: AppEvent) -> Self {
        Self {
            state: ButtonState::Normal,
            label: String::from(label),
            theme: BLUE,
            event_sender,
            on_press,
        }
    }
    pub fn is_pressed(&self) -> bool {
        self.state == ButtonState::Active
    }
    pub const fn theme(mut self, theme: Theme) -> Self {
        self.theme = theme;
        self
    }
    const fn colors(&self) -> (Color, Color, Color, Color) {
        let theme = self.theme;
        match self.state {
            ButtonState::Normal => (theme.background, theme.text, theme.shadow, theme.highlight),
            ButtonState::Selected => (theme.highlight, theme.text, theme.shadow, theme.highlight),
            ButtonState::Active => (theme.background, theme.text, theme.highlight, theme.shadow),
        }
    }
}

impl Widget for Button {
    fn handle_event(&mut self, event: AppEvent) {
        match event {
            AppEvent::Clear(_) => self.state = ButtonState::Normal,
            AppEvent::NoFocus => self.state = ButtonState::Normal,
            AppEvent::Focus => self.state = ButtonState::Selected,
            AppEvent::KeyEvent(key) => match key.kind {
                KeyEventKind::Press => {
                    if self.state == ButtonState::Selected {
                        self.state = ButtonState::Active;
                        self.event_sender.send(self.on_press.clone().into());
                    }
                }
                KeyEventKind::Release => {
                    if self.state == ButtonState::Active {
                        self.state = ButtonState::Selected
                    }
                }
                _ => {}
            },
            _ => {}
        }
    }
    fn draw(&self, area: Rect, buf: &mut Buffer, _: &mut Option<u16>) {
        let (background, text, shadow, _highlight) = self.colors();
        let block = Block::bordered()
            .border_type(ratatui::widgets::BorderType::Rounded)
            .style(Style::new().fg(shadow));
        let inner = block.inner(area);
        block.render(area, buf);
        buf.set_style(inner, Style::new().bg(background).fg(text));
        /*
                // render top line if there's enough space
                if area.height > 2 {
                    buf.set_string(
                        area.x,
                        area.y,
                        "▔".repeat(area.width as usize),
                        Style::new().fg(highlight).bg(background),
                    );
                }
                // render bottom line if there's enough space
                if area.height > 1 {
                    buf.set_string(
                        area.x,
                        area.y + area.height - 1,
                        "▁".repeat(area.width as usize),
                        Style::new().fg(shadow).bg(background),
                    );
                }
        */
        let line: Line<'_> = String::from(&self.label).into();
        // render label centered
        buf.set_line(
            area.x + (area.width.saturating_sub(line.width() as u16)) / 2,
            area.y + (area.height.saturating_sub(1)) / 2,
            &line,
            area.width,
        );
    }
}
