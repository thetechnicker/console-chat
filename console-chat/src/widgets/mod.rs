use crate::event::AppEvent;
//use ratatui::widgets as ratatui_widgets;
//use ratatui::crossterm::event::KeyEvent;
use ratatui::{buffer::Buffer, layout::Rect};
use std::fmt::Debug;

pub trait Widget: Debug {
    fn handle_event(&mut self, event: AppEvent);
    fn draw(&self, area: Rect, buf: &mut Buffer, ret: &mut Option<u16>);
}

pub mod input;
pub use input::*;
pub mod button;
pub use button::*;
