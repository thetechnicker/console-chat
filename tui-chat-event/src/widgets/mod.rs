use crate::event::WidgetEvent;
//use ratatui::widgets as ratatui_widgets;
//use ratatui::crossterm::event::KeyEvent;
use std::fmt::Debug;

pub trait Widget: Debug {
    fn handle_event(&mut self, event: WidgetEvent);
}

pub mod input;
pub use input::*;
