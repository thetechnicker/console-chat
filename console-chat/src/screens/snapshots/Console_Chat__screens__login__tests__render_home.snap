---
source: src/screens/login.rs
expression: terminal.backend()
---
"╔══════════════════════════════════════════════════════════════════════════════╗"
"║                                                                              ║"
"║                                                                              ║"
"║                                                                              ║"
"║            ╭────────────────────────────────────────────────────╮            ║"
"║            │Username                                            │            ║"
"║            ╰────────────────────────────────────────────────────╯            ║"
"║            ╭────────────────────────────────────────────────────╮            ║"
"║            │Password                                            │            ║"
"║            ╰────────────────────────────────────────────────────╯            ║"
"║            ╭─────────────────────────╮╭─────────────────────────╮            ║"
"║            │          Login          ││         Anonym          │            ║"
"║            ╰─────────────────────────╯╰─────────────────────────╯            ║"
"║            ╭────────────────────────────────────────────────────╮            ║"
"║            │                        Exit                        │            ║"
"║            ╰────────────────────────────────────────────────────╯            ║"
"║                                                                              ║"
"║                                                                              ║"
"║                                                                              ║"
"╚══════════════════════════════════════════════════════════════════════════════╝"
